### PR TITLE
encoder self-attn should mask out invalid pixels when used as queries…

### PR DESCRIPTION
**Pull request by atiorh**
For all transformer encoder self-attention layers, this repo currently masks out the pixels that were added via zero-padding (applied for batching different sized images) using the `key_padding_mask` argument of `nn.MultiheadAttention`. However, the same zero-padding generated pixels are used as queries for encoder self-attention as well and we compute non-zero attention outputs for these pixels (because the queries locations corresponding to these invalid pixels are not masked).
This PR proposes to use the 3-D `attn_mask` argument of `nn.MultiheadAttention` of shape `(batch_size * n_heads, query_seqlen, key_seqlen)` and mask out these invalid pixels when used as queries as well as keys.

This is not necessarily a bug but I wanted to understand whether the authors have already considered this in which case I am more than happy to understand why my way of thinking is wrong/not beneficial for the attention mechanism.

Thanks for the awesome repo! 🙏

### CODE REVIEW

**Comment by atiorh** [comment]
I just tested this and masking with `float('-inf')` for queries results in NaN due to softmax over vectors full of infs. That can be fixed by implementing the masking in another way so at this point, I just want to understand what the authors think about query masking :)

**Comment by alcinos** [comment]
Not sure I understand the issue exactly, but from what I understand you seem concerned by the fact that padded pixels may attend to whatever they want in the image during self-attention.
Since the values of these pixels are completely ignored the rest of the time (FFNs are element-wise, and in the attention we block the other pixels from attending padded pixels), it doesn't matter what value they take.
Let me know what you think.

**Comment by atiorh** [comment]
Thanks for the explanation! Let me clarify my concern: the current version of encoder self-attention blocks all queries from attending zero-padded pixels as keys (via `key_padding_mask`) which is desirable. However, it also treats the queries corresponding to zero-padded pixels in the same way as valid pixel locations and lets them attend over the same set of valid keys. Is this desirable? My concern is that the queries corresponding to these zero-padded pixels will compute spurious attention patterns and interfere with training/learning dynamics.

Additional remarks:
- The value embeddings corresponding to these pixel locations are highly likely to include non-zero entries due to convolutions in the backbone so the output of self-attention for these queries is also highly likely to be non-zero.
- Given that the sequence that enters self-attention is already downsampled heavily, the effect of this may be suppressed.

This may be a minute detail but I still wanted to be on the same page and appreciated your follow up :)

**Comment by alcinos** [comment]
Sure, the element in the sequence corresponding to padded pixels can attend to whatever they want, and will contain whatever garbage value. Now, what I don't understand is how this will "interfere with training/learning dynamics". These elements won't be used in any way further down in the model (since they are masked out from "valid" elements), so they won't even receive gradients.

**Comment by atiorh** [comment]
Okay that makes perfect sense, the garbage values do not contribute to any valid pixel. Thanks for the discussion 👍
